pmd-core/src/main/java/net/sourceforge/pmd/util/log/AntLogHandler.java

```diff
@@ -23,10 +25,52 @@
     private Project project;
 
     private static final Formatter FORMATTER = new PmdLogFormatter();
+    private static final Level[] LOG_LEVELS = {
+        Level.SEVERE, Level.WARNING, Level.INFO, Level.CONFIG, Level.FINEST,
+    };
 
     public AntLogHandler(Project project) {
         this.project = project;
     }
+    
+    public Level getAntLogLevel() {
+        Class<?> clazz;
+        int maxLevel = Project.MSG_ERR;
+        for (final BuildListener l : project.getBuildListeners()) {
+            clazz = l.getClass();
+            Field declaredField = null;
+            try {
+                while (BuildListener.class.isAssignableFrom(clazz)) {
+                    try {
+                        declaredField = clazz.getDeclaredField("msgOutputLevel");
+                    } catch (final NoSuchFieldException ignored) {
+                        try {
+                            declaredField = clazz.getDeclaredField("logLevel");
```

Comment:
Instead of trying the second field as fallback, we could also explicitly check the type (is assignable / instanceof) for the listeners we know:

* org.apache.tools.ant.DefaultLogger and org.apache.tools.ant.XmlLogger : here the field is called "msgOutputLevel" - if we fail to find this field, then Ant changed and we will need to update PMD...
* org.apache.tools.ant.taskdefs.RecorderEntry : the field is called "loglevel"

You should be able then to say - avoiding to search the hierarchy:

```java
int logLevel = Project.MSG_INFO;
Field logLevelField = null;
if (l instanceof DefaultLogger) {
    logLevelField = DefaultLogger.class.getDeclaredField("msgOutputLevel");
} else if (l instanceof XmlLogger) {
    logLevelField = XmlLogger.class.getDeclaredField("msgOutputLevel");
} else if (l instanceof RecorderEntry) {
    logLevelField = RecorderEntry.class.getDeclaredField("loglevel");
} else {
    syserr?? LOG.warn("Unsupported build listener: " + l.getClass());
}
if (logLevel != null) {
    logLevelField.setAccessible(true);
    logLevel = logLevelField.getInt(firstListener);
}
```


I don't think, that we can speculate of other BuildListener implementations - if they subclass from DefaultLogger (and many obviously do), we support them - and they probably reuse msgOutputLevel, since it is protected. If someone creates an own listener, we don't know anything and could only guess (logLevel, level, msgOutputLevel, ??).

I'd suggest, that in case we can't determine the loglevel, we should simply use the default level MSG_INFO and log a warning, that we don't know, how to determine the current loglevel from listener class XXX (maybe with the suggestion to subclass DefaultLogger).

Reply:
Actually this second case was added to handle the Ant JUnit Rule, as it doesn't even implement `BuildLogger` and uses a separate field in camel-case as seen here. I may be missing more cases non the less given what you say.

Reply:
Can you give me a pointer? Is it the Ant JUnit Task? Or something else?

Reply:
It's the `org.apache.tools.ant.BuildFileRule$AntTestListener` configured by the `@Rule BuildFileRule` in JUnit tests.

Reply:
Interesting... this `AntTestListener` is even a private class. So maybe, we could just check for the declared field `logLevel` on the given class `l.getClass()`. That way, we can even write a unit test to verify the logging - we currently use BuldFileRule ourselves at e.g. https://github.com/pmd/pmd/blob/master/pmd-core/src/test/java/net/sourceforge/pmd/ant/PMDTaskTest.java